### ISSUE TITLE
query arangodb is slow 

### ISSUE THREAD

**Issue by volinh** [arangodb/arangodb#10496]
## My Environment

* __ArangoDB Version__:        3.5.1
* __Storage Engine__:             RocksDB 
* __Deployment Mode__:       Single Server 
* __Deployment Strategy__:    ArangoDB Starter DCOS -->
* __Configuration__:               default
* __Infrastructure__:               own 
* __Operating System__:        centos7>
* __Total RAM in your machine__:        125GB
* __Disks in use__:        HDD
* __Used Package__:             

## Component, Query & Data

__Affected feature__:
AQL query using web interface | java driver 

__AQL query (if applicable)__:
FOR doc IN domains
    FILTER doc.domain IN [a.com, b.com]
    RETURN doc._key

__Dataset__:
3 billion records , one record combine 4 field 

__Size of your Dataset on disk__:
125GB

__Problem__:
i create hash index in domain field, but it still slow, the result has about  1 million to 2 million records , time run query up to 2000s 

__Expected result__:
are there any way to speed up query ?

**Comment by DiscoPYF** [comment]
Hi volinh,

Did you try using the "Explain" and "Profile" button when writing your query in the web interface to make sure the AQL optimizer make use of the index?

I think it may be useful for the ArangoDB team to see the output of these commands. These commands also exist in arangosh: https://www.arangodb.com/docs/stable/aql/execution-and-performance-explaining-queries.html

**Comment by volinh** [comment]
![image](https://user-images.githubusercontent.com/25150446/69403788-10d0d680-0d2e-11ea-9b34-5ae486418580.png)
hi DiscoPYF,
this is plan of query

**Comment by Simran-B** [comment]
Your index has a pretty low selectivity and you return the full documents. This means that a lot of data has to be read from disk, which could be slow as you use an HDD and no flash storage. Do you actually need the full documents? If yes, and taking the large amount of data into account which needs to be sent over the network to the client, consider using a `stream`ing query to get the first cursor batch faster:

https://www.arangodb.com/docs/stable/aql/invocation-with-arangosh.html#setting-options
https://www.arangodb.com/docs/stable/http/aql-query-cursor-accessing-cursors.html

**Comment by volinh** [comment]
dear Simran-B,
i test speed of reading and writing of server
read: 600mb/s
write: 500mb/s

i create a another collection , it has 2 million documents , filed hostname has only 2 value, selectivity is 0%, but the time to load 1 million full documents from the server down through the web interface or java driver is 40 seconds. 

![image](https://user-images.githubusercontent.com/25150446/69520354-05c8b100-0f8f-11ea-99b0-d6163c4f40a8.png)
![image](https://user-images.githubusercontent.com/25150446/69520387-1f69f880-0f8f-11ea-8d31-6ef058eafa70.png)

i think there are problems in hashing index in arangodb when size of collection is too large

**Comment by Simran-B** [comment]
Do you actually retrieve all documents in your test taking 40 seconds? The web interface limits queries to 1000 results by default. In case of the Java driver, do you exhaust the cursor entirely?

**Comment by volinh** [comment]
yes ,the collection has 2 million documents and i retrieve 1 miliion documents by java driver

![image](https://user-images.githubusercontent.com/25150446/69537507-da57bd80-0fb2-11ea-86c0-889cff366673.png)

**Comment by Simran-B** [comment]
What is the performance difference if you run the exact query twice, but one time with the hash index in place and another time with that index removed? Is it faster without the index?

**Comment by volinh** [comment]
no, Indexing still helps the query faster. 
I mean why hash indexing is so slow with with a 3 billion documents collection ? ,
i see in document the hash index is O(1)

**Comment by Simran-B** [comment]
Hash indexes are **amortized** O(1), but this doesn't tell you much about the practical performance, especially in cases of low selectivity. They work best if each value is unique. Moreover, RocksDB does not use classic hash tables. The index properties are closer to a skiplist index. It is a sorted index, which means you can read ranges. All indexes are also persistent if you use RocksDB as storage engine. It doesn't actually matter whether you create a `hash`, `skiplist` or `persistent` index using RocksDB. The internal index type is the same under the hood (and we dropped the former two in the Web UI to reflect this). An in-memory hash index which you get when you use the MMFiles storage engine is faster for point lookups, but volatile.

The question is if the index is I/O bound for you or if you run into a different limitation which is hardware related (note that RocksDB is optimized for fast storage) or whether there is an inefficiency on ArangoDB's side. How much CPU, RAM and HDD utilization do you see during the query execution?
